File: tests/polyglot-pilot/csharp/src/ProductionApp.cs
// ProductionApp.cs — toy signup feature for the C# polyglot pilot.
//
// Minimal composition root: in-memory registry + audit log, exposes
// SignupAsync (driving operation) and CaptureUniverseAsync (state inspection
// at port-exposed slots).
//
// Universe slots exposed:
//   - "registry.users" : IReadOnlyList<UserRecord>
//   - "audit.events"   : IReadOnlyList<AuditEvent>
//
// Internal field names are NOT part of the universe — refactoring internals
// stays GREEN.

#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace NWave.Polyglot.SignupPilot;

public sealed record UserRecord(string Email);

public sealed record AuditEvent(string Type, string Email);

public sealed class DuplicateSignupException : Exception
{
    public string Email { get; }

    public DuplicateSignupException(string email)
        : base($"Duplicate signup rejected: {email}")
    {
        Email = email;
    }
}

public sealed class ProductionApp
{
    // Driven-port state (in-memory).
    private readonly List<UserRecord> _users = new();
    private readonly List<AuditEvent> _events = new();

    /// <summary>
    /// Driving port — signup a user by email. Rejects duplicates by throwing
    /// DuplicateSignupException. On success: appends to registry AND appends
    /// a single UserSignedUp audit event.
    /// </summary>
    public Task<UserRecord> SignupAsync(string email)
    {
        var normalized = email.Trim().ToLowerInvariant();
        if (normalized.Length == 0)
        {
            throw new ArgumentException("signup: email must be non-empty", nameof(email));
        }
        if (_users.Any(u => u.Email == normalized))
        {
            throw new DuplicateSignupException(normalized);
        }
        var record = new UserRecord(normalized);
        _users.Add(record);
        _events.Add(new AuditEvent("UserSignedUp", normalized));
        return Task.FromResult(record);
    }

    /// <summary>
    /// State-inspection port — return a snapshot of the universe slots
    /// requested. Snapshot returns defensive copies so test assertions cannot
    /// mutate production state by accident. Anonymous-shaped dictionaries
    /// match the test's expected predicates (StateDelta uses reflection
    /// deep-equal on public properties).
    /// </summary>
    public Task<Dictionary<string, object?>> CaptureUniverseAsync(IEnumerable<string> keys)
    {
        var snapshot = new Dictionary<string, object?>();
        foreach (var key in keys)
        {
            switch (key)
            {
                case "registry.users":
                    snapshot[key] = _users
                        .Select(u => (object?)new Dictionary<string, object?> { ["Email"] = u.Email })
                        .ToList();
                    break;
                case "audit.events":
                    snapshot[key] = _events
                        .Select(e => (object?)new Dictionary<string, object?>
                        {
                            ["Type"] = e.Type,
                            ["Email"] = e.Email,
                        })
                        .ToList();
                    break;
                default:
                    // Unknown slot — record null so state-delta sees the absence
                    // explicitly rather than silently fabricating a value.
                    snapshot[key] = null;
                    break;
            }
        }
        return Task.FromResult(snapshot);
    }
}

File: tests/polyglot-pilot/java/src/main/java/pilot/ProductionApp.java
/*
 * Toy ProductionApp — minimal signup feature for the polyglot pilot.
 *
 * Composition root: constructs the in-memory registry and audit log, exposes
 * `signup` (driving operation) and `captureUniverse` (state inspection at
 * port-exposed slots). Real-feature shape: in-process domain + driven ports
 * (here both in-memory).
 *
 * Universe slots exposed:
 *   - "registry.users" : List<Map<String,Object>> with {"email": String}
 *   - "audit.events"   : List<Map<String,Object>> with {"type": String, "email": String}
 *
 * Internal field names are NOT part of the universe — refactoring internals
 * stays GREEN.
 */
package pilot;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ProductionApp {

    /** Thrown when a signup is attempted with an email already in the registry. */
    public static final class DuplicateSignupException extends RuntimeException {
        private static final long serialVersionUID = 1L;
        public final String email;

        public DuplicateSignupException(String email) {
            super("Duplicate signup rejected: " + email);
            this.email = email;
        }
    }

    // Driven-port state (in-memory).
    private final List<Map<String, Object>> users = new ArrayList<>();
    private final List<Map<String, Object>> events = new ArrayList<>();

    public ProductionApp() {
        // Toy feature — nothing to wire.
    }

    /**
     * Driving port — signup a user by email. Rejects duplicates by throwing
     * {@link DuplicateSignupException}. On success: appends to registry AND
     * appends a single {@code UserSignedUp} audit event.
     */
    public Map<String, Object> signup(String rawEmail) {
        if (rawEmail == null) {
            throw new IllegalArgumentException("signup: email must be non-null");
        }
        String email = rawEmail.trim().toLowerCase();
        if (email.isEmpty()) {
            throw new IllegalArgumentException("signup: email must be non-empty");
        }
        for (Map<String, Object> u : users) {
            if (email.equals(u.get("email"))) {
                throw new DuplicateSignupException(email);
            }
        }
        Map<String, Object> record = new LinkedHashMap<>();
        record.put("email", email);
        users.add(record);

        Map<String, Object> event = new LinkedHashMap<>();
        event.put("type", "UserSignedUp");
        event.put("email", email);
        events.add(event);

        return new LinkedHashMap<>(record);
    }

    /**
     * State-inspection port — return a snapshot of the universe slots requested.
     * Snapshot returns deep-copied entries so test assertions cannot mutate
     * production state by accident.
     */
    public Map<String, Object> captureUniverse(Set<String> keys) {
        Map<String, Object> snapshot = new LinkedHashMap<>();
        for (String key : keys) {
            switch (key) {
                case "registry.users" -> snapshot.put(key, deepCopyList(users));
                case "audit.events" -> snapshot.put(key, deepCopyList(events));
                default -> snapshot.put(key, null);
            }
        }
        return snapshot;
    }

    private static List<Map<String, Object>> deepCopyList(List<Map<String, Object>> src) {
        List<Map<String, Object>> out = new ArrayList<>(src.size());
        for (Map<String, Object> item : src) {
            out.add(new LinkedHashMap<>(item));
        }
        return out;
    }
}
